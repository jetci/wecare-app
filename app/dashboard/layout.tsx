'use client';

import { AuthGuard } from '@/components/auth/AuthGuard';
import { Sidebar, SidebarMobile } from '@/components/layout/Sidebar';
import Header from '@/components/layout/Header';
import React, { useState } from 'react';

type Props = { children: React.ReactNode };

const LoadingScreen = () => (
  <div className="flex h-screen items-center justify-center bg-gray-100">
    <p>Loading...</p>
  </div>
);

export default function DashboardLayout({ children }: Props) {
  const [sidebarOpen, setSidebarOpen] = useState(false);

  return (
    <AuthGuard>
      <div className="min-h-screen bg-gray-100">
        <SidebarMobile sidebarOpen={sidebarOpen} setSidebarOpen={setSidebarOpen} />
        <Sidebar />
        <div className="lg:pl-72">
          <Header setSidebarOpen={setSidebarOpen} />
          <main className="py-10">
            <div className="px-4 sm:px-6 lg:px-8">{children}</div>
          </main>
        </div>
      </div>
    </AuthGuard>
  );
}
