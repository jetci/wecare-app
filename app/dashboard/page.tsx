"use client";

import { useEffect } from 'react';
import { useAuth } from '@/context/AuthContext';
import { useRouter } from 'next/navigation';

export default function DashboardPage() {
  const { loading, isAuthenticated, role } = useAuth();
  const router = useRouter();

  useEffect(() => {
    if (!loading) {
      if (!isAuthenticated) {
        router.push('/login');
      } else {
        switch (role) {
          case 'COMMUNITY':
            router.replace('/dashboard/community');
            break;
          case 'DRIVER':
            router.replace('/dashboard/driver');
            break;
          case 'HEALTH_OFFICER':
            router.replace('/dashboard/health-officer');
            break;
          case 'EXECUTIVE':
            router.replace('/dashboard/executive');
            break;
          case 'ADMIN':
            router.replace('/dashboard/admin');
            break;
          case 'DEVELOPER':
            router.replace('/dashboard/developer');
            break;
          default:
            router.replace('/dashboard');
        }
      }
    }
  }, [loading, isAuthenticated, role, router]);

  return <p>กำลังไปยังหน้า Dashboard ตามบทบาท...</p>;
}
