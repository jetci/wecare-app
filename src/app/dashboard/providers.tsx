'use client';
import { AuthProvider } from '@/context/AuthContext';

export function DashboardProviders({ children }: { children: React.ReactNode }) {
  return <AuthProvider>{children}</AuthProvider>;
}
