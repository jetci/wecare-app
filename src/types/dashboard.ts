// src/types/dashboard.ts

import type { ComponentType } from 'react';

export interface CacheStats {
  hits: number
  misses: number
  keysCount: number
}

export interface JobStats {
  successCount: number
  failedCount: number
  items: unknown[] // TODO: define Job type
}

export interface DeploymentStats {
  pendingCount: number
  successCount: number
  failureCount: number
  lastDeployed: string // ISO timestamp
}

export interface HealthPoint {
  timestamp: string // ISO timestamp
  cpuUsage: number
  memoryUsage: number
  dbConnections: number
}

export interface SystemHealth {
  cpuUsage: number
  memoryUsage: number
  dbConnections: number
}

export interface Notification {
  id: string
  message: string
  timestamp: string // ISO timestamp
}

export interface Community {
  id: string
  name: string
  total: number
  inCare: number
  transferred: number
}

export interface Patient {
  id: string
  name: string
  status: string
}

// Dashboard menu item type
export interface DashboardMenuItem {
  key: string
  title: string
  path: string
  icon?: ComponentType<unknown>
  children?: DashboardMenuItem[]
}

export interface DashboardMenus {
  community: DashboardMenuItem[]
  driver: DashboardMenuItem[]
  healthOfficer: DashboardMenuItem[]
  executive: DashboardMenuItem[]
  admin: DashboardMenuItem[]
}

// Menu structure configuration placeholder
// define dashboardMenus in a separate file

export interface Patient {
  id: string
  name: string
  status: string
}