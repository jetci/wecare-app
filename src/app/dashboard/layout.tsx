"use client";

import React, { ReactNode, useEffect } from "react";
import Link from "next/link";
import { useRouter } from "next/navigation";
import useSWR from "swr";
import Header from "@/components/layout/Header";
import { Role } from "@/types/roles";

const fetcher = (url: string) =>
  fetch(url, {
    headers: { Authorization: `Bearer ${localStorage.getItem("accessToken")}` },
  }).then((res) => res.json());

const allDashboards = [
  { key: "community", title: "ประชาชน", path: "/dashboard/community" },
  { key: "driver", title: "เจ้าหน้าที่กู้ชีพ/คนขับรถ", path: "/dashboard/driver" },
  { key: "health-officer", title: "เจ้าหน้าที่สาธารณสุข", path: "/dashboard/health-officer" },
  { key: "executive", title: "ผู้บริหาร", path: "/dashboard/executive" },
  { key: "admin", title: "ผู้ดูแลระบบ", path: "/dashboard/admin" }
];

interface Props {
  children: ReactNode;
  role: string;
}

export default function DashboardLayout({ children, role }: Props) {
  const router = useRouter();
  const { data: profileRaw } = useSWR("/api/auth/profile", fetcher);
  const profile = profileRaw?.user ?? profileRaw;
  const isDevUser = profile?.nationalId === "3500200461028";

  // Sidebar items
  let sidebarItems;
  if (role === 'community') {
    // Community users see only community dashboard
    sidebarItems = allDashboards.filter((m) => m.key === role);
  } else if (isDevUser) {
    // Dev user sees all dashboards
    sidebarItems = allDashboards;
  } else {
    // Other users see their own dashboard only
    sidebarItems = allDashboards.filter((m) => m.key === role);
  }

  // inactivity logout
  useEffect(() => {
    const events = ["click", "keydown", "mousemove", "scroll"];
    const updateActivity = () =>
      localStorage.setItem("lastActivity", Date.now().toString());
    events.forEach((e) => window.addEventListener(e, updateActivity));
    const interval = setInterval(() => {
      const last = parseInt(localStorage.getItem("lastActivity") || "0", 10);
      if (Date.now() - last > 120 * 60 * 1000) router.push("/login");
    }, 60 * 1000);
    return () => {
      events.forEach((e) => window.removeEventListener(e, updateActivity));
      clearInterval(interval);
    };
  }, [router]);

  return (
    <div className="flex flex-col h-screen">
      <Header />
      <div className="flex flex-1">
        <aside className="w-64 bg-gray-800 text-white p-4">
          <h2 className="text-xl font-bold mb-4">WeCare Dev Nav</h2>
          <nav className="space-y-2">
            {sidebarItems.map((item) => (
              <Link
                key={item.key}
                href={item.path}
                className="block p-2 hover:bg-gray-700 rounded"
              >
                {item.title}
              </Link>
            ))}
          </nav>
          <button
            onClick={() => {
              localStorage.removeItem("accessToken");
              document.cookie =
                "refreshToken=; path=/; secure; SameSite=Lax; max-age=0";
              router.push("/login");
            }}
            className="block mt-4 hover:underline"
          >
            ออกจากระบบ
          </button>
        </aside>
        <main className="flex-1 overflow-auto bg-gray-100 p-6">
          {children}
        </main>
      </div>
    </div>
  );
}