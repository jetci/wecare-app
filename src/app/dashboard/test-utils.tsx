import React, { type ReactNode } from 'react';
import {
  render as rtlRender,
  screen,
  fireEvent,
  waitFor,
  act,
  within,
  waitForElementToBeRemoved,
  cleanup,
  type RenderOptions,
} from '@testing-library/react';
// Removed renderHook import; hook testing should use render from @testing-library/react or custom wrappers
import { vi } from 'vitest';
import RoleGuard from '@/components/RoleGuard';
import { AuthProvider } from '@/context/AuthContext';
import { Role } from '@/types/roles';

// Mock next/navigation
vi.mock('next/navigation', () => ({
  useRouter: () => ({
    push: (to: string) => window.history.pushState({}, '', to),
    replace: (to: string) => window.history.pushState({}, '', to),
    prefetch: vi.fn(),
    refresh: vi.fn(),
    asPath: window.location.pathname,
  }),
  usePathname: () => window.location.pathname,
  useSearchParams: () => new URLSearchParams(window.location.search),
}));

// Mock Google Maps API
vi.mock('@react-google-maps/api', () => ({
  GoogleMap: (props: any) => <div data-testid="mock-google-map" {...props} />,
  Marker: (props: any) => <div data-testid="mock-google-marker" {...props} />,
  useJsApiLoader: () => ({ isLoaded: true, loadError: null }),
}));

interface ProvidersProps {
  children: ReactNode;
  route?: string;
  role?: string;
}

function AllProviders({ children, route = '/', role = 'ADMIN' }: ProvidersProps) {
  window.history.pushState({}, '', route);
  const roleEnum = (Role as any)[role.toUpperCase() as keyof typeof Role] as Role;
  return (
    <AuthProvider>
      <RoleGuard allowedRoles={[roleEnum]}>{children}</RoleGuard>
    </AuthProvider>
  );
}

function renderWithProviders(
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'wrapper'> & { route?: string; role?: string }
) {
  return rtlRender(ui, {
    wrapper: ({ children }) => <AllProviders {...options}>{children}</AllProviders>,
  });
}

export {
  renderWithProviders,
  renderWithProviders as render,
  screen,
  fireEvent,
  waitFor,
  act,
  within,
  waitForElementToBeRemoved,
  cleanup,
  // renderHook removed,
};
