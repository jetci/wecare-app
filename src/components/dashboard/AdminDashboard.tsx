import React, { useState } from 'react'
import useSWR from 'swr'
import { User, LogEntry, ApiKey, ConfirmInfo, UnknownObject } from '@/types/components'
import { ConfirmationModal } from '../ui/ConfirmationModal'
import { SummaryCard } from '../ui/SummaryCard'
import toast from 'react-hot-toast'

const fetcher = (url: string) => fetch(url).then(res => res.json())
const tabs = [
  'Users', 'Logs', 'Settings', 'API Keys', 'Maintenance', 'Thresholds'
]

const AdminDashboard: React.FC = () => {
  const [activeTab, setActiveTab] = useState<string>(tabs[0])
  const [confirm, setConfirm] = useState<ConfirmInfo | null>(null)

  // Data
  const { data: users, error: usersError, mutate: mutateUsers } = useSWR<User[]>('/api/admin/users', fetcher)
  const { data: logs, error: logsError } = useSWR<LogEntry[]>('/api/admin/logs', fetcher)
  const { data: settings, error: settingsError } = useSWR('/api/admin/settings', fetcher)
  const { data: apiKeys, error: keysError, mutate: mutateKeys } = useSWR<ApiKey[]>('/api/admin/api-keys', fetcher)
  const { data: thresholds, error: thrError, mutate: mutateThr } = useSWR<UnknownObject[]>('/api/admin/settings/thresholds', fetcher)

  // Actions
  const handleAction = async (url: string, method='PUT', body?: UnknownObject, refresh?: () => void) => {
    try {
      const res = await fetch(url, { method, headers: {'Content-Type':'application/json'}, body: body && JSON.stringify(body) })
      if (!res.ok) throw new Error('Action failed')
      toast.success('สำเร็จ')
      refresh && refresh()
    } catch (err) {
      toast.error('ล้มเหลว')
    }
  }

  return (
    <div className="p-4">
      {/* Tabs */}
      <div className="flex space-x-4 border-b mb-4">
        {tabs.map(tab => (
          <button key={tab}
            className={`px-4 py-2 ${activeTab===tab ? 'border-b-2 border-blue-500' : ''}`}
            onClick={() => setActiveTab(tab)}>{tab}</button>
        ))}
      </div>
      <div>
        {activeTab === 'Users' && (
          <div>
            {!users && <p>Loading users...</p>}
            {usersError && <p className="text-red-500">Error loading users</p>}
            {users?.map((u: User) => (
              <div key={u.id} className="flex justify-between p-2 border-b">
                <span>{u.firstName} {u.lastName} ({u.role})</span>
                <div className="space-x-2">
                  <button className="px-2 py-1 bg-green-500 text-white rounded"
                    onClick={() => handleAction(`/api/admin/users/${u.id}/approve`, 'PUT', undefined, mutateUsers)}>
                    Approve
                  </button>
                  <button className="px-2 py-1 bg-red-500 text-white rounded"
                    onClick={() => handleAction(`/api/admin/users/${u.id}/disable`, 'PUT', undefined, mutateUsers)}>
                    Disable
                  </button>
                  <button className="px-2 py-1 bg-yellow-500 text-white rounded"
                    onClick={() => handleAction(`/api/admin/users/${u.id}/reset-password`, 'POST')}>
                    Reset PW
                  </button>
                </div>
              </div>
            ))}
          </div>
        )}
        {activeTab === 'Logs' && (
          <div>
            {!logs && <p>Loading logs...</p>}
            {logsError && <p className="text-red-500">Error loading logs</p>}
            <div className="max-h-64 overflow-auto">
              {logs?.map((log: LogEntry, i: number) => (
                <div key={i} className="text-sm border-b p-1">
                  [{log.timestamp}] {log.user} {log.action} from {log.ip}
                </div>
              ))}
            </div>
          </div>
        )}
        {activeTab === 'Settings' && (
          <div>
            {/* System Settings form */}
            {!settings && <p>Loading settings...</p>}
            {settings && <pre>{JSON.stringify(settings,null,2)}</pre>}
          </div>
        )}
        {activeTab === 'API Keys' && (
          <div>
            <button className="mb-2 px-3 py-1 bg-blue-500 text-white rounded"
              onClick={() => handleAction('/api/admin/api-keys','POST',undefined, mutateKeys)}>
              Create Key
            </button>
            {keysError && <p className="text-red-500">Error</p>}
            {apiKeys?.map((k: ApiKey) => (
              <div key={k.id} className="flex justify-between p-1 border-b">
                <span>{k.key}</span>
                <button className="px-2 py-1 bg-red-500 text-white rounded"
                  onClick={() => handleAction(`/api/admin/api-keys/${k.id}`,'DELETE',undefined, mutateKeys)}>
                  Revoke
                </button>
              </div>
            ))}
          </div>
        )}
        {activeTab === 'Maintenance' && (
          <div className="space-y-2">
            <button className="px-4 py-2 bg-blue-500 text-white rounded"
              onClick={() => handleAction('/api/admin/backup','POST')}>Backup Data</button>
            <button className="px-4 py-2 bg-gray-500 text-white rounded"
              onClick={() => handleAction('/api/admin/cache/clear','POST')}>Clear Cache</button>
          </div>
        )}
        {activeTab === 'Thresholds' && (
          <div>
            {!thresholds && <p>Loading...</p>}
            {thresholds && <pre>{JSON.stringify(thresholds,null,2)}</pre>}
          </div>
        )}
      </div>
      {confirm && (
        <ConfirmationModal
          open
          title="Confirm Action"
          message={confirm.message}
          onConfirm={() => { confirm.onConfirm(); setConfirm(null) }}
          onCancel={() => setConfirm(null)}
        />
      )}
    </div>
  )
}

export default AdminDashboard
