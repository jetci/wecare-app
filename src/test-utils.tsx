// src/test-utils.tsx
import React from 'react'
import { render as rtlRender, RenderOptions } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { AuthProvider } from '@/context/AuthContext'
import RoleGuard from '@/components/RoleGuard'
import { Role } from '@/types/roles'

export function render(ui: React.ReactElement, options?: RenderOptions) {
  return rtlRender(
    <AuthProvider>
      <RoleGuard allowedRoles={[Role.COMMUNITY, Role.ADMIN, Role.DEVELOPER]}>{ui}</RoleGuard>
    </AuthProvider>,
    options
  )
}

export * from '@testing-library/react'
export { userEvent }
