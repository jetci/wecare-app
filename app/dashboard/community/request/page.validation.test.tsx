import '@testing-library/jest-dom/vitest';
import React from 'react';
import { render, screen, fireEvent, waitFor } from '@/app/dashboard/test-utils';
import CommunityRequestPage from './page';
import { vi } from 'vitest';

// Mock useCreateRide hook
const createRideMock = vi.fn().mockResolvedValue(undefined);
vi.mock('@/hooks/useCreateRide', () => ({
  useCreateRide: () => ({ createRide: createRideMock }),
}));

describe('CommunityRequestPage Validation', () => {
  beforeEach(() => {
    createRideMock.mockClear();
  });

  it('shows validation errors when submitting empty form', async () => {
    render(<CommunityRequestPage />, { role: 'COMMUNITY', route: '/dashboard/community/request' });
    fireEvent.click(screen.getByRole('button', { name: /send/i }));
    expect(await screen.findByText(/Origin is required/)).toBeInTheDocument();
    expect(screen.getByText(/Description is required/)).toBeInTheDocument();
  });

  it('submits form with valid data', async () => {
    render(<CommunityRequestPage />, { role: 'COMMUNITY', route: '/dashboard/community/request' });
    fireEvent.input(screen.getByLabelText(/Origin/i), { target: { value: 'Home' } });
    fireEvent.input(screen.getByLabelText(/รายละเอียด/i), { target: { value: 'Need help' } });
    fireEvent.click(screen.getByRole('button', { name: /send/i }));
    await waitFor(() => expect(createRideMock).toHaveBeenCalledWith({ origin: 'Home', description: 'Need help' }));
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<CommunityRequestPage />, { role: 'COMMUNITY', route: '/dashboard/community/request' });
    expect(asFragment()).toMatchSnapshot();
  });
});

