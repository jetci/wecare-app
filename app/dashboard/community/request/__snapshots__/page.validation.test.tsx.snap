// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CommunityRequestPage Validation > matches snapshot 1`] = `<DocumentFragment />`;
